app/components/app_activity_log_component.rb

```diff
@@ -302,60 +302,105 @@ def programmes_by_id
 
   def decision_expiration_events
     all_programmes = Programme.all.to_a
+    events = []
 
-    AcademicYear.all.filter_map do |academic_year|
+    AcademicYear.find_each do |academic_year|
       next if academic_year >= AcademicYear.current
 
-      vaccinated_programmes =
-        all_programmes.select do |programme|
+      not_vaccinated_programmes =
+        all_programmes.reject do |programme|
           patient.vaccination_status(programme:, academic_year:).vaccinated?
         end
 
-      programmes_with_expired_consents =
-        consents
-          .select { it.academic_year == academic_year }
-          .flat_map { programmes_for(it) }
-          .reject { vaccinated_programmes.include?(it) }
-
-      programmes_with_expired_triages =
-        triages
-          .select { it.academic_year == academic_year }
-          .flat_map { programmes_for(it) }
-          .reject { vaccinated_programmes.include?(it) }
-
-      programmes_with_expired_psds =
-        patient_specific_directions
-          .select { it.academic_year == academic_year }
-          .flat_map { programmes_for(it) }
-          .reject { vaccinated_programmes.include?(it) }
-
-      expired_items = []
-      if programmes_with_expired_consents.any?
-        expired_items += ["consent", "health information"]
-      end
-      expired_items << "triage outcome" if programmes_with_expired_triages.any?
-      expired_items << "PSD status" if programmes_with_expired_psds.any?
+      vaccinated_but_seasonal_programmes =
+        all_programmes.select do |programme|
+          patient.vaccination_status(programme:, academic_year:).vaccinated? &&
+            programme.seasonal?
+        end
 
-      next if expired_items.empty?
+      expired_items = {
+        vaccinated_but_seasonal:
+          get_expired_items(
+            academic_year:,
+            programmes: vaccinated_but_seasonal_programmes
+          ),
+        not_vaccinated:
+          get_expired_items(
+            academic_year:,
+            programmes: not_vaccinated_programmes
+          )
+      }
 
-      programmes_with_expired_items = [
-        programmes_with_expired_consents,
-        programmes_with_expired_triages,
-        programmes_with_expired_psds
-      ].flatten.uniq
+      %i[not_vaccinated vaccinated_but_seasonal].each do |category|
```

Comment:
Rather than needing to iterate over the keys and then look them up later, I think we could do something like:

```rb
expired_items.each do |category, category_expired_items|
  # ...
end
```